envoy/cilium_host_map.h

```diff
@@ -1,19 +1,49 @@
 #pragma once
 
+#include <arpa/inet.h>
+
 #include "envoy/local_info/local_info.h"
 #include "envoy/upstream/cluster_manager.h"
 #include "envoy/event/dispatcher.h"
 
 #include "common/common/logger.h"
+#include "common/network/utility.h"
 #include "envoy/config/subscription.h"
 #include "envoy/singleton/instance.h"
 #include "envoy/thread_local/thread_local.h"
 
 #include "cilium/nphds.pb.h"
 
+#include "absl/numeric/int128.h"
+
+// std::hash specialization for Abseil uint128, needed for unordered_map key.
+namespace std {
+  template <> struct hash<absl::uint128>
+  {
+    size_t operator()(const absl::uint128& x) const
+    {
+      return hash<uint64_t>{}(absl::Uint128Low64(x)) ^ (hash<uint64_t>{}(absl::Uint128High64(x)) << 1);
+    }
+  };
+}
+
 namespace Envoy {
 namespace Cilium {
 
+template <typename I> I ntoh(I);
+template <> inline uint32_t ntoh(uint32_t addr) { return ntohl(addr); }
+template <> inline absl::uint128 ntoh(absl::uint128 addr) { return Network::Utility::Ip6ntohl(addr); }
+template <typename I> I hton(I);
+template <> inline uint32_t hton(uint32_t addr) { return htonl(addr); }
+template <> inline absl::uint128 hton(absl::uint128 addr) { return Network::Utility::Ip6htonl(addr); }
+
+template <typename I> I masked(I addr, unsigned int plen) {
+  const unsigned int PLEN_MAX = sizeof(I)*8;
+  return addr & ~hton((I(1) << (PLEN_MAX - plen)) - 1);
+};
+
+enum ID : uint64_t { UNKNOWN = 0, HOST = 1, WORLD = 2, CLUSTER = 3 };
```

Comment:
Why do you need to have those defined in the filter? I only see `ID:WORLD` used.
You're supposed to receive the mapping for "host" through NPHDS. I think it should be the case for "cluster" too.
OK with defining constants for "unknown" and "world", but I don't think you should define an enum.

Reply:
Alternatively we could make the "not found" case return value ID configurable in the filter config?

Reply:
IMO it would be cleanest to require the default CIDR to be configured in the host map itself and simply return 0 for the "not found" case.

Reply:
OK, I removed the unused values (HOST, CLUSTER) from the enum for now. The default return for the "not found" case is still ID::WORLD (2), but now CIDRs with 0 prefix lengths are supported. So when Cilium daemon is changed to pass in "0.0.0.0/0" and "::/0" CIDRs we can remove the WORLD definition and simply return 0 when an address cannot be mapped to any ID.